File: DriftGame/uracer-desktop/src/com/bitfire/uracer/screen/TransitionFactory.java
package com.bitfire.uracer.screen;

import com.badlogic.gdx.utils.LongMap;
import com.bitfire.uracer.screen.transitions.CrossFader;
import com.bitfire.uracer.screen.transitions.Fader;
import com.bitfire.utils.Hash;

public final class TransitionFactory {

    private static final LongMap<ScreenTransition> transitions = new LongMap<ScreenTransition>();
    private static ScreenFactory screenFactory = null;

    private TransitionFactory() {
    }

    public static void init(ScreenFactory factory) {
        screenFactory = factory;
    }

    public static ScreenTransition getTransition(TransitionType transitionType) {
        ScreenTransition transition = transitions.get(transitionType.hash);

        if (transition == null) {
            transition = createTransition(transitionType);
            transitions.put(transitionType.hash, transition);
        } else {
            transition.reset();
        }

        return transition;
    }

    private static ScreenTransition createTransition(TransitionType transitionType) {
        switch (transitionType) {
            case CrossFader:
                return new CrossFader(screenFactory);
            case Fader:
                return new Fader(screenFactory);
            default:
            case None:
                return null;
        }
    }

    public static void dispose() {
        for (ScreenTransition t : transitions.values()) {
            t.dispose();
        }
    }

    public enum TransitionType {
        None, CrossFader, Fader;

        public long hash;

        TransitionType() {
            hash = Hash.APHash(this.name());
        }
    }
}

File: DriftGame/uracer-desktop/src/com/bitfire/uracer/screen/TransitionFactory.kt
package com.bitfire.uracer.screen

import com.badlogic.gdx.utils.LongMap
import com.bitfire.uracer.screen.transitions.CrossFader
import com.bitfire.uracer.screen.transitions.Fader
import com.bitfire.utils.Hash

object TransitionFactory {

    private val transitions = LongMap<ScreenTransition>()
    private lateinit var screenFactory: ScreenFactory

    @JvmStatic
    fun init(factory: ScreenFactory) {
        screenFactory = factory
    }

    @JvmStatic
    fun getTransition(transitionType: TransitionType): ScreenTransition {
        var transition = transitions.get(transitionType.hash)

        if (transition == null) {
            transition = createTransition(transitionType)
            transitions.put(transitionType.hash, transition)
        } else {
            transition.reset()
        }

        return transition
    }

    private fun createTransition(transitionType: TransitionType) = when (transitionType) {
        TransitionType.CrossFader -> CrossFader(screenFactory)
        TransitionType.Fader -> Fader(screenFactory)
        TransitionType.None -> null
    }

    @JvmStatic
    fun dispose() {
        for (transition in transitions.values()) transition.dispose()
    }

    enum class TransitionType {
        None, CrossFader, Fader;

        var hash = Hash.apHash(this.name)
    }
}
